### PR TITLE
Database

### LINKED ISSUE #14: Set up DB.

### LINKED ISSUE #29: Create a Database classe that uses firebase and allow to answer, post and disply questions

### PR DESCRIPTION
**Pull request by RomainLogean**
Implementation of FirebaseDatabaseAdapter.
Here is the firebase database : https://console.firebase.google.com/u/0/project/epforuml-38150/database/epforuml-38150-default-rtdb/data

Closes #14, Closes #29

### CODE REVIEW

File: app/src/main/java/com/github/ybecker/epforuml/database/FirebaseDatabaseAdapter.kt [hunk 1]
```diff
@@ -1,59 +1,345 @@
...
+    private val questionIdPath = "questionId"
+    private val answerIdPath = "answerId"
+
+    private val courseNamePath = "courseName"
+    private val usernamePath = "username"
+
+    private val questionTextPath = "questionText"
+    private val answerTextPath = "answerText"
 
     override fun availableCourses(): Set<Course> {
```
Comment (Caro1290) [inline app/src/main/java/com/github/ybecker/epforuml/database/FirebaseDatabaseAdapter.kt:34]: There is some code repetition through the different functions:  getQuestionsForCourse, availableCourses, getQuestionAnswers, getUserQuestions and getUserAnswers have the same structure
Maybe it would be better for exemple to do one single function that recognizes the different paths.

File: app/src/main/java/com/github/ybecker/epforuml/database/FirebaseDatabaseAdapter.kt [hunk 2]
```diff
@@ -1,59 +1,345 @@
...
+            return getUser
+        }
+
+        // create a space for the new question in sb and save its id
+        val newUser = User(userId, username, emptyList(), emptyList())
+        db.child(usersPath).child(userId).setValue(newUser)
+        return newUser
     }
 
     override fun getQuestionById(id: String): Question? {
```
Comment (Caro1290) [inline app/src/main/java/com/github/ybecker/epforuml/database/FirebaseDatabaseAdapter.kt:194]: Same remark here for the functions getQuestionById(id: String), getAnswerById(id: String), getUserById(id: String) and getCourseById(id: String)

File: app/src/main/java/com/github/ybecker/epforuml/database/FirebaseDatabaseAdapter.kt [hunk 3]
```diff
@@ -1,59 +1,345 @@
...
+    private val answerTextPath = "answerText"
 
     override fun availableCourses(): Set<Course> {
-        TODO("Not yet implemented")
+        val future = CompletableFuture<Set<Course>>()
+        // go in "courses" dir
+        db.child(coursesPath).get().addOnSuccessListener {
+            val courses = mutableSetOf<Course>()
+            // add every course that in not null in "courses" in the map
+            for (courseSnapshot in it.children) {
```
Comment (Caro1290) [inline app/src/main/java/com/github/ybecker/epforuml/database/FirebaseDatabaseAdapter.kt:40]: You should test if (it.value == null)

File: app/src/main/java/com/github/ybecker/epforuml/database/FirebaseDatabaseAdapter.kt [hunk 4]
```diff
@@ -1,59 +1,345 @@
...
 
-    override fun getAnswersForQuestion(question: Question): Set<Answer> {
-        TODO("Not yet implemented")
+    //Note that using question.answers in the main is false because you don't take new values in the db into account !
+    override fun getQuestionAnswers(question: Question): Set<Answer> {
+        val future = CompletableFuture<Set<Answer>>()
+        // go in "question/questionId" dir
+        db.child(questionsPath).child(question.questionId).child(answersPath).get().addOnSuccessListener {
+            val answers = mutableSetOf<Answer>()
+            // add every question's answer that is not null in the map
```
Comment (Caro1290) [inline app/src/main/java/com/github/ybecker/epforuml/database/FirebaseDatabaseAdapter.kt:82]: You should test if (it.value == null)

File: app/src/main/java/com/github/ybecker/epforuml/database/FirebaseDatabaseAdapter.kt [hunk 5]
```diff
@@ -1,59 +1,345 @@
...
+        return future.get(5,TimeUnit.SECONDS)
+    }
+
+    //Note that using user.question in the main is false because you don't take new values in the db into account !
+    override fun getUserQuestions(user: User): Set<Question> {
+        val future = CompletableFuture<Set<Question>>()
+        // go in "user/userId" dir
+        db.child(usersPath).child(user.userId).child(questionsPath).get().addOnSuccessListener {
+            val questions = mutableSetOf<Question>()
+            // add every user's question that is not null in the map
```
Comment (Caro1290) [inline app/src/main/java/com/github/ybecker/epforuml/database/FirebaseDatabaseAdapter.kt:103]: You should test if (it.value == null)

File: app/src/main/java/com/github/ybecker/epforuml/database/FirebaseDatabaseAdapter.kt [hunk 6]
```diff
@@ -1,59 +1,345 @@
...
+
+        return future.get(5,TimeUnit.SECONDS)
+    }
+
+    override fun getUserAnswers(user: User): Set<Answer> {
+        val future = CompletableFuture<Set<Answer>>()
+        // go in "user/userId" dir
+        db.child(usersPath).child(user.userId).child(answersPath).get().addOnSuccessListener {
+            val answers = mutableSetOf<Answer>()
+            // add every user's question that is not null in the map
```
Comment (Caro1290) [inline app/src/main/java/com/github/ybecker/epforuml/database/FirebaseDatabaseAdapter.kt:123]: You should test if (it.value == null)

File: app/src/main/java/com/github/ybecker/epforuml/database/FirebaseDatabaseAdapter.kt [hunk 7]
```diff
@@ -1,59 +1,345 @@
...
+                if (answer != null) {
+                    answers.add(answer)
+                }
+            }
+            future.complete(answers)
+        }.addOnFailureListener {
+            future.completeExceptionally(it)
+        }
+
+        return future.get(5,TimeUnit.SECONDS)
```
Comment (Caro1290) [inline app/src/main/java/com/github/ybecker/epforuml/database/FirebaseDatabaseAdapter.kt:135]: Maybe it would be a good thing to declare a constante for the "5" because it is used in several functions

**Review by Caro1290 — APPROVED** [review]
Great work on this code!! 

The part you've worked on is absolutely fondamental to the overall project, and I can see that you've put a lot of effort into it. I think that it is good to underline that it is a really important piece of work, and I appreciate all the work that have been done regarding the creation of the structure of the code. 
There are some repetitive sections of code that could potentially be refactored, and the use of an emulator for the tests would certainly be a real advantage.
Also dont forget to add the test on "it" (if it is null or not) in the different functions I mentioned.
However the overall quality of the code and the architecture, the way the different classes are build is really smart. 

Thank you for your hard work!
